src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -69,19 +66,151 @@ public int compare(Event e1, Event e2) {
             queryResult.add(currentRange);
         }
 
-        return queryResult;
+        Collection<TimeRange> optionalCheck = considerOptional(
+            request.getDuration());
+
+        if (optionalCheck.isEmpty()) {
+            return queryResult;
+        }
+
+        return optionalCheck;
+    }
+
+    private Collection<TimeRange> considerOptional(long duration) {
+        ArrayList<TimeRange> optionalResult = new ArrayList<TimeRange>();
+
+        if (queryResult.isEmpty()) {
+            return optionalResult;
+        }
+
+        if (guestEventQueue.isEmpty()) {
+            return optionalResult;
+        }
+
+        Iterator<TimeRange> availableRangeIterator = queryResult.iterator();
+        TimeRange workingTimeRange = availableRangeIterator.next();
+        TimeRange currentEventTimeRange = guestEventQueue.poll().getWhen();
+
+        eventIteration:
+        while (true) {
+
+            if (workingTimeRange.start() >= currentEventTimeRange.end()) {
+                if (guestEventQueue.isEmpty()) {
+                    break eventIteration;
+                }
+
+                currentEventTimeRange = guestEventQueue.poll().getWhen();
+            }
+
+            while (!currentEventTimeRange.overlaps(workingTimeRange)
+                    && workingTimeRange.start() < currentEventTimeRange.end()) {
+                if (!availableRangeIterator.hasNext()) {
+                    // Might be able too immediately return here.
+                    break eventIteration;
+                }
+
+                optionalResult.add(workingTimeRange);
+                workingTimeRange = availableRangeIterator.next();
+            }
+
+            if (currentEventTimeRange.overlaps(workingTimeRange)) {
+                TimeRange[] rangeCut = handleOverlap(
+                    currentEventTimeRange, workingTimeRange, duration);
+
+                if (rangeCut[0] != null) {
```

Comment:
nit:May consider introduce a new type with an indicator if it is optional range.